// storybook-addon-next-router/dist/storybook-addon-next-router.esm.js
import { action } from '@storybook/addon-actions';
import { makeDecorator } from '@storybook/addons';
import React from 'react';
import Router from 'next/router';
import { RouterContext } from 'next/dist/next-server/lib/router-context';

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var withNextRouter = /*#__PURE__*/makeDecorator({
  name: 'NextRouter',
  parameterName: 'nextRouter',
  wrapper: function wrapper(getStory, context, settings) {
    Router.router = _extends({
      route: '/',
      pathname: '/',
      query: {},
      asPath: '/',
      push: function push() {
        action('nextRouter.push').apply(void 0, arguments);
        return Promise.resolve(true);
      },
      replace: function replace() {
        action('nextRouter.replace').apply(void 0, arguments);
        return Promise.resolve(true);
      },
      reload: function reload() {
        action('nextRouter.reload').apply(void 0, arguments);
      },
      back: function back() {
        action('nextRouter.back').apply(void 0, arguments);
      },
      prefetch: function prefetch() {
        action('nextRouter.prefetch').apply(void 0, arguments);
        return Promise.resolve();
      },
      beforePopState: function beforePopState() {
        action('nextRouter.beforePopState').apply(void 0, arguments);
      },
      events: {
        on: function on() {
          action('nextRouter.events.on').apply(void 0, arguments);
        },
        off: function off() {
          action('nextRouter.events.off').apply(void 0, arguments);
        },
        emit: function emit() {
          action('nextRouter.events.emit').apply(void 0, arguments);
        }
      },
      isFallback: false
    }, settings.options, settings.parameters);
    return React.createElement(RouterContext.Provider, {
      value: Router.router
    }, getStory(context));
  }
});

export { withNextRouter };
