// storybook-addon-next-router/dist/storybook-addon-next-router.cjs.development.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var addonActions = require('@storybook/addon-actions');
var addons = require('@storybook/addons');
var React = _interopDefault(require('react'));
var Router = _interopDefault(require('next/router'));
var routerContext = require('next/dist/next-server/lib/router-context');

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

var withNextRouter = /*#__PURE__*/addons.makeDecorator({
  name: 'NextRouter',
  parameterName: 'nextRouter',
  wrapper: function wrapper(getStory, context, settings) {
    Router.router = _extends({
      route: '/',
      pathname: '/',
      query: {},
      asPath: '/',
      push: function push() {
        addonActions.action('nextRouter.push').apply(void 0, arguments);
        return Promise.resolve(true);
      },
      replace: function replace() {
        addonActions.action('nextRouter.replace').apply(void 0, arguments);
        return Promise.resolve(true);
      },
      reload: function reload() {
        addonActions.action('nextRouter.reload').apply(void 0, arguments);
      },
      back: function back() {
        addonActions.action('nextRouter.back').apply(void 0, arguments);
      },
      prefetch: function prefetch() {
        addonActions.action('nextRouter.prefetch').apply(void 0, arguments);
        return Promise.resolve();
      },
      beforePopState: function beforePopState() {
        addonActions.action('nextRouter.beforePopState').apply(void 0, arguments);
      },
      events: {
        on: function on() {
          addonActions.action('nextRouter.events.on').apply(void 0, arguments);
        },
        off: function off() {
          addonActions.action('nextRouter.events.off').apply(void 0, arguments);
        },
        emit: function emit() {
          addonActions.action('nextRouter.events.emit').apply(void 0, arguments);
        }
      },
      isFallback: false
    }, settings.options, settings.parameters);
    return React.createElement(routerContext.RouterContext.Provider, {
      value: Router.router
    }, getStory(context));
  }
});

exports.withNextRouter = withNextRouter;
