// @nrwl/next/src/generators/application/lib/add-linting.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addLinting = void 0;
const tslib_1 = require("tslib");
const linter_1 = require("@nrwl/linter");
const devkit_1 = require("@nrwl/devkit");
const react_1 = require("@nrwl/react");
const run_tasks_in_serial_1 = require("@nrwl/workspace/src/utilities/run-tasks-in-serial");
function addLinting(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const lintTask = yield linter_1.lintProjectGenerator(host, {
            linter: options.linter,
            project: options.projectName,
            tsConfigPaths: [
                devkit_1.joinPathFragments(options.appProjectRoot, 'tsconfig.app.json'),
            ],
            eslintFilePatterns: [`${options.appProjectRoot}/**/*.{ts,tsx,js,jsx}`],
            skipFormat: true,
        });
        if (options.linter === linter_1.Linter.EsLint) {
            const reactEslintJson = react_1.createReactEslintJson(options.appProjectRoot);
            devkit_1.updateJson(host, devkit_1.joinPathFragments(options.appProjectRoot, '.eslintrc.json'), () => {
                var _a, _b;
                if ((_b = (_a = reactEslintJson.overrides) === null || _a === void 0 ? void 0 : _a[0].parserOptions) === null || _b === void 0 ? void 0 : _b.project) {
                    reactEslintJson.overrides[0].parserOptions.project = [
                        `${options.appProjectRoot}/tsconfig(.*)?.json`,
                    ];
                }
                return reactEslintJson;
            });
        }
        const installTask = devkit_1.addDependenciesToPackageJson(host, react_1.extraEslintDependencies.dependencies, react_1.extraEslintDependencies.devDependencies);
        return run_tasks_in_serial_1.runTasksInSerial(lintTask, installTask);
    });
}
exports.addLinting = addLinting;
