// @nrwl/workspace/src/core/hasher/hasher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Hasher = void 0;
const tslib_1 = require("tslib");
const fs_1 = require("fs");
const file_utils_1 = require("../file-utils");
const child_process_1 = require("child_process");
const file_hasher_1 = require("./file-hasher");
const hashing_impl_1 = require("./hashing-impl");
const minimatch = require("minimatch");
const perf_hooks_1 = require("perf_hooks");
const stripJsonComments = require("strip-json-comments");
class Hasher {
    constructor(projectGraph, nxJson, options, hashing = undefined) {
        this.projectGraph = projectGraph;
        this.nxJson = nxJson;
        this.options = options;
        if (!hashing) {
            this.hashing = hashing_impl_1.defaultHashing;
            this.fileHasher = file_hasher_1.defaultFileHasher;
        }
        else {
            // this is only used for testing
            this.hashing = hashing;
            this.fileHasher = new file_hasher_1.FileHasher(hashing);
            this.fileHasher.clear();
        }
        this.projectHashes = new ProjectHasher(this.projectGraph, this.hashing);
    }
    hashTasks(tasks) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            perf_hooks_1.performance.mark('hasher:hash:start');
            const r = yield Promise.all(tasks.map((t) => this.hash(t)));
            perf_hooks_1.performance.mark('hasher:hash:end');
            perf_hooks_1.performance.measure('hasher:hash', 'hasher:hash:start', 'hasher:hash:end');
            return r;
        });
    }
    hash(task) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            const command = this.hashing.hashArray([
                task.target.project || '',
                task.target.target || '',
                task.target.configuration || '',
                JSON.stringify(task.overrides),
            ]);
            const values = (yield Promise.all([
                this.projectHashes.hashProject(task.target.project, [
                    task.target.project,
                ]),
                this.implicitDepsHash(),
                this.runtimeInputsHash(),
            ]));
            const value = this.hashing.hashArray([
                Hasher.version,
                command,
                ...values.map((v) => v.value),
            ]);
            return {
                value,
                details: {
                    command,
                    sources: values[0].sources,
                    implicitDeps: values[1].sources,
                    runtime: values[2].runtime,
                },
            };
        });
    }
    runtimeInputsHash() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.runtimeInputs)
                return this.runtimeInputs;
            perf_hooks_1.performance.mark('hasher:runtime inputs hash:start');
            const inputs = this.options && this.options.runtimeCacheInputs
                ? this.options.runtimeCacheInputs
                : [];
            if (inputs.length > 0) {
                try {
                    const values = (yield Promise.all(inputs.map((input) => new Promise((res, rej) => {
                        child_process_1.exec(input, (err, stdout, stderr) => {
                            if (err) {
                                rej(err);
                            }
                            else {
                                res({ input, value: `${stdout}${stderr}`.trim() });
                            }
                        });
                    }))));
                    const value = this.hashing.hashArray(values.map((v) => v.value));
                    const runtime = values.reduce((m, c) => ((m[c.input] = c.value), m), {});
                    perf_hooks_1.performance.mark('hasher:runtime inputs hash:end');
                    perf_hooks_1.performance.measure('hasher:runtime inputs hash', 'hasher:runtime inputs hash:start', 'hasher:runtime inputs hash:end');
                    return { value, runtime };
                }
                catch (e) {
                    throw new Error(`Nx failed to execute runtimeCacheInputs defined in nx.json failed:\n${e.message}`);
                }
            }
            else {
                this.runtimeInputs = Promise.resolve({ value: '', runtime: {} });
            }
            return this.runtimeInputs;
        });
    }
    implicitDepsHash() {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (this.implicitDependencies)
                return this.implicitDependencies;
            perf_hooks_1.performance.mark('hasher:implicit deps hash:start');
            const implicitDeps = Object.keys(this.nxJson.implicitDependencies || {});
            const filesWithoutPatterns = implicitDeps.filter((p) => p.indexOf('*') === -1);
            const patterns = implicitDeps.filter((p) => p.indexOf('*') !== -1);
            const implicitDepsFromPatterns = patterns.length > 0
                ? (this.projectGraph.allWorkspaceFiles || [])
                    .filter((f) => !!patterns.find((pattern) => minimatch(f.file, pattern)))
                    .map((f) => f.file)
                : [];
            const fileNames = [
                ...filesWithoutPatterns,
                ...implicitDepsFromPatterns,
                //TODO: vsavkin move the special cases into explicit ts support
                'tsconfig.base.json',
                'package-lock.json',
                'yarn.lock',
                'pnpm-lock.yaml',
            ];
            this.implicitDependencies = Promise.resolve().then(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const fileHashes = [
                    ...fileNames.map((file) => {
                        const hash = this.fileHasher.hashFile(file);
                        return { file, hash };
                    }),
                    ...this.hashGlobalConfig(),
                ];
                const combinedHash = this.hashing.hashArray(fileHashes.map((v) => v.hash));
                perf_hooks_1.performance.mark('hasher:implicit deps hash:end');
                perf_hooks_1.performance.measure('hasher:implicit deps hash', 'hasher:implicit deps hash:start', 'hasher:implicit deps hash:end');
                return {
                    value: combinedHash,
                    sources: fileHashes.reduce((m, c) => ((m[c.file] = c.hash), m), {}),
                };
            }));
            return this.implicitDependencies;
        });
    }
    hashGlobalConfig() {
        return [
            {
                hash: this.fileHasher.hashFile('nx.json', (file) => {
                    try {
                        const r = JSON.parse(stripJsonComments(file));
                        delete r.projects;
                        return JSON.stringify(r);
                    }
                    catch (e) {
                        return '';
                    }
                }),
                file: 'nx.json',
            },
        ];
    }
}
exports.Hasher = Hasher;
Hasher.version = '1.0';
class ProjectHasher {
    constructor(projectGraph, hashing) {
        this.projectGraph = projectGraph;
        this.hashing = hashing;
        this.sourceHashes = {};
        this.workspaceJson = this.readConfigFile(file_utils_1.workspaceFileName());
        this.nxJson = this.readConfigFile('nx.json');
    }
    hashProject(projectName, visited) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            return Promise.resolve().then(() => tslib_1.__awaiter(this, void 0, void 0, function* () {
                const deps = this.projectGraph.dependencies[projectName] || [];
                const depHashes = (yield Promise.all(deps.map((d) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    if (visited.indexOf(d.target) > -1) {
                        return null;
                    }
                    else {
                        visited.push(d.target);
                        return yield this.hashProject(d.target, visited);
                    }
                })))).filter((r) => !!r);
                const projectHash = yield this.hashProjectNodeSource(projectName);
                const sources = depHashes.reduce((m, c) => {
                    return Object.assign(Object.assign({}, m), c.sources);
                }, { [projectName]: projectHash });
                const value = this.hashing.hashArray([
                    ...depHashes.map((d) => d.value),
                    projectHash,
                ]);
                return { value, sources };
            }));
        });
    }
    hashProjectNodeSource(projectName) {
        return tslib_1.__awaiter(this, void 0, void 0, function* () {
            if (!this.sourceHashes[projectName]) {
                this.sourceHashes[projectName] = new Promise((res) => tslib_1.__awaiter(this, void 0, void 0, function* () {
                    const p = this.projectGraph.nodes[projectName];
                    const fileNames = p.data.files.map((f) => f.file);
                    const values = p.data.files.map((f) => f.hash);
                    const workspaceJson = JSON.stringify(this.workspaceJson.projects[projectName] || '');
                    const nxJson = JSON.stringify(this.nxJson.projects[projectName] || '');
                    res(this.hashing.hashArray([
                        ...fileNames,
                        ...values,
                        workspaceJson,
                        nxJson,
                    ]));
                }));
            }
            return this.sourceHashes[projectName];
        });
    }
    readConfigFile(name) {
        try {
            const res = JSON.parse(stripJsonComments(fs_1.readFileSync(name).toString()));
            if (!res.projects)
                res.projects = {};
            return res;
        }
        catch (e) {
            return { projects: {} };
        }
    }
}
